Guide for enabling HTTPS/TLS (#9536)
keycloak/keycloak: docs/guides/src/main/server/enabletls.adoc

<@kc.start parameters="--https-protocols=<protocol>[,<protocol>]"/>

To also allow TLSv1.2, use for example `kc.sh start --http-protocols=TLSv1.3,TLSv1.2`.

== Switch the HTTPS port
Keycloak listens for HTTPS traffic on port `8443` by default. To change this port, use:
<@kc.start parameters="--https-port=<port>"/>

== Using a truststore
Keycloak uses a truststore to store certificates to verify clients that are communicating with Keycloak, e.g. to use mutualTLS, certificate-bound tokens or X.509 authentication. This truststore is used for outgoing https requests as well as mTLS requests.

You can configure the location of this truststore by running:
<@kc.start parameters="--https-trust-store-file=/path/to/file"/>

=== Setting the truststore password
You can set a secure password for your truststore using the `https.trust-store.password` option:
<@kc.start parameters="--https.trust-store.password=<value>"/>
If no password is set, the default password `password` is used.

> == Securing credentials

Author: just a placeholder for now.

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>

<@tmpl.guide

> title="Configure Keycloak to use TLS"

Reviewer: I'd probably avoid `Keycloak` in titles, as it's kinda obvious isn't it? Can leave this as is for now, then we can review this when we have a bit more guides in, and have a follow-up PR to review all titles separately.

Author: +1 for the second iteration for polishing.

Author: will change it nevertheless while i'm on it

> summary="Learn how to use HTTPS/TLS with Keycloak"

Reviewer: Would expand on the summary to mention that it includes configuring Keycloak's https certificates as well as truststore for outgoing requests and mTLS.

includedOptions="https.* http.enabled">

Transport Layer Security (short: TLS) is crucial to exchange data over a secured channel. For production environments, you should never expose Keycloak endpoints through HTTP, as sensitive data is at the core of what Keycloak exchanges with other applications. In this guide you will learn how to configure Keycloak to use HTTPS/TLS.

== Configure TLS in Keycloak

> Keycloak can be configured to load the needed certificate infrastructure either from a certificate / key file pair in PEM format or from an existing keystore file. When both alternatives are set up, the certificate / key file configuration takes precedence over the keystore configuration.

Reviewer: "certificate / key file pair in PEM format" is a bit long and hard to read. Perhaps something like:

```suggestion
Keycloak can be configured to load the needed certificate infrastructure using files in PEM format or from a Java KeyStore. When both alternatives are set up, the PEM files takes precedence over the Java KeyStores.
```

Author: will change with the upcoming version 👍

> === Providing a certificate / key file pair

Reviewer: ```suggestion
=== Providing certificates in PEM format
```

When you use a pair of matching certificate / private key files in PEM format, configure Keycloak to use them by running:

> <@kc.all parameters="--https-certificate-file=/path/to/certfile.pem --https-certificate-key-file=/path/to/keyfile.pem"/>

Reviewer: I'd suggest replacing `@kc.all` with `@kc.start` throughout

Keycloak creates a keystore out of these files in memory and uses this keystore afterwards.

> === Providing a keystore

Reviewer: ```suggestion
=== Providing a Java KeyStore
```

As an alternative, you can use an existing keystore by running:
<@kc.all parameters="--https-key-store-file=/path/to/existing-keystore-file"/>

> When no keystore file is configured explicitly, but `http.enabled` is set to false, Keycloak tries to use a `server.keystore` file in the `conf` location by default.

Reviewer: I'd move this paragraph above the "As an alternative", and instead of "tries to use" say something like "looks for the default "server.keystore" file.

==== Setting a keystore password
You can set a secure password for your keystore using the `https.key-store.password` option:
<@kc.all parameters="--https-key-store-password=<value>"/>

> [NOTE]

Reviewer: Perhaps we should rather add a section at the end with something like "Securing secrets", which can link to the "handling secrets" guide in the future?

We recommend to not set the password in plaintext via the CLI or in `conf/keycloak.properties`, but instead using good practices such as using a vault or an environment variable - both can be substituted in the properties file. If no password is set, the default password `password` is used.

> == Enable other TLS Version(s)

Reviewer: ```suggestion
== Configuring TLS protocols
```

> As per default, Keycloak uses the current TLS Version, which is 1.3 at the time of writing. If you need to use an older version, you may change the version by running:

Reviewer: ```suggestion
By default, Keycloak does not enable deprecated TLS protocols. In situations where clients only support deprecated protocols, first consider upgrading the client, but as a temporary work-around you can enable deprecated protocols with:
```

> <@kc.all parameters="--https-protocols=<version>[,<version>]"/>

Reviewer: ```suggestion
<@kc.all parameters="--https-protocols=<protocol>[,<protocol>]"/>
```

To also allow TLSv1.2, you can use for example `kc.sh start --http-protocols=TLSv1.3,TLSv1.2`.

== Switch the HTTPS port

> You can switch the port Keycloak listens on for HTTPS traffic by running:

Reviewer: Perhaps this should mention the default port used?

<@kc.all parameters="--https-port=<port>"/>

== Using a truststore

> Keycloak uses a truststore to store certificates to verify clients that are communicating with Keycloak, e.g. to use mutualTLS, certificate-bound tokens or X.509 authentication.

Reviewer: Should also mention that the truststore is used both for outgoing https requests, as well as mTLS requests.

You can configure the location of this truststore by running:
<@kc.all parameters="--https-trust-store-file=/path/to/file"/>

=== Setting a truststore password
As with the keystore, you can (and should) set a secure password for your truststore. This password can be configured using the `https.trust-store.password` option:
<@kc.all parameters="--https-trust-store-password=<value>"/>

> [NOTE]

Reviewer: Same as above, I suggest having a separate section for handling secrets.

== Configure TLS in Keycloak
Keycloak can be configured to load the needed certificate infrastructure either from a certificate / key file pair in PEM format or from an existing keystore file. When both alternatives are set up, the certificate / key file configuration takes precedence over the keystore configuration.

=== Providing a certificate / key file pair
When you use a pair of matching certificate / private key files in PEM format, configure Keycloak to use them by running:

<@kc.all parameters="--https-certificate-file=/path/to/certfile.pem --https-certificate-key-file=/path/to/keyfile.pem"/>

Keycloak creates a keystore out of these files in memory and uses this keystore afterwards.

=== Providing a keystore
As an alternative, you can use an existing keystore by running:
<@kc.all parameters="--https-key-store-file=/path/to/existing-keystore-file"/>

When no keystore file is configured explicitly, but http.enabled is set to false, Keycloak will try to use a `server.keystore` file in the `conf` subdirectory by default.

[NOTE]

> No matter which alternative you choose for setting up the TLS configuration, please make sure to set a secure password for your keystore using the `https.key-store.password` option. We recommend to not set the password in plaintext via CLI or in `conf/keycloak.properties`, but instead using good practices such as using a vault or an environment variable - both can be substituted in the properties file. If no password is set, the default password `password` is used.

Reviewer 2: Why do we need to set a password when using PEM files?

Author: that was a guess, but I got an answer from the quarkus team here: https://github.com/quarkusio/quarkus/discussions/22851#discussioncomment-1961299 so yes, vert.x stores it internally in a keystore in memory, so the question is now: should we consider this "safe enough" to not advocate for using a safe password? Fine either way, tbh :)

Reviewer 2: I do not think there is two ways, but no need to set a password if using PEM files.

Author: Updated, can you pls check again?

== Enable other TLS Version(s)
As per default, Keycloak uses the current TLS Version, which is 1.3 at the time of writing. If you need to use an older version, you may change the version by running:

<@kc.all parameters="--https-protocols=<version>[,<version>]"/>

To also allow TLSv1.2, you can use for example `kc.sh start --http-protocols=TLSv1.3,TLSv1.2`.

== Switch the HTTPS port
You can switch the port Keycloak listens on for HTTPS traffic by running:
<@kc.all parameters="--https-port=<port>"/>

> == Using a truststore

Reviewer 2: I'm not sure if this part is accurate.

The trust-store configuration here should be about setting up the HTTPS listener to verify client certificates.

The examples you are giving here are related to another level of configuration. Please, see https://www.keycloak.org/docs/latest/server_installation/#_truststore.

Author: also updated, pls check again.
